### ISSUE TITLE
Can a dev review my Rust bindings?

### ISSUE THREAD

**Issue by adamkewley** [ebiggers/libdeflate#55]
Hi,

I've been using `libdeflate` at work and I really like it. As a side-project--and to practice unsafe + FFI coding in Rust--I have implemented bindings to most of the `libdeflate` API but, if there's a dev around, could you review my design? I'm gradually getting closer to rolling it out publicly (it's on cargo, but I haven't put it on any social media sites etc.)

- Repo: https://github.com/adamkewley/libdeflater
- API docs: https://docs.rs/libdeflater/0.1.1/libdeflater/
- Example gz compression: https://github.com/adamkewley/libdeflater/blob/master/examples/gz_compress.rs
- Example gz decompression: https://github.com/adamkewley/libdeflater/blob/master/examples/gz_decompress.rs

The benchmarks show that `libdeflate` does very well when compared with Rust's most popular compression library (`flate2`), but the corpus is mostly small files (favourable to `libdeflate`), I'm not comparing the compression ratio at this default setting, and I might be using `flate2` sub-optimally, so take the numbers with a pinch of salt.

My main goal with the crate is to have a well-documented API so that developers can add `libdeflate` into their codebase easily, but also to not try and hide the API behind (e.g.) a streaming interface, so that developers *know* how the underlying protocol works (therefore making the decision to use `libdeflate` a deliberate one)

**Comment by ebiggers** [comment]
I took a quick look, and it looks sane.

Maybe `Compressor::new()` should pass failures to the caller rather than panicing, given that a NULL return from `libdeflate_alloc_compressor()` can mean that the compression level is invalid?  I.e. it doesn't always mean "out of memory".

Is the omission of bindings for the `*_decompress_ex()` functions (which have the `actual_in_nbytes_ret` argument) intentional?  Likewise for the Adler32 and CRC-32 functions.

Also a few typos I happened to notice:
* panic message in `decompress_deflate()` has wrong function name
* panic message in `Compressor::new()` has wrong function name
* comment above `compress_zlib()` says `out_deflate_data` instead of `out_zlib_data`
* comment above `compress_gzip()` says `out_deflate_data` instead of `out_gzip_data`

**Comment by adamkewley** [comment]
Thanks for this @ebiggers , it is very valuable to have another dev look at this.

To deal with the fact that a NULL might be thrown by `libdeflate_alloc_compressor()` because the compression level is invalid, I moved returning a result into `CompressionLvl::new(i32)`, so the signature for that is now: `pub fn new(level: i32) -> CompressionLevelResult`. This means that library users should not be able to construct an arbitrary `CompressionLvl` without the number being range-checked. 

I decided to add the `Result` at this point, rather than in `::new` to make it clear, from a type-system POV, that the only thing a library user can do erroneously is provide an invalid compression level. Logically, there are no other side-effects to allocating a compressor other than memory allocation, and returning a `Result` implies that there are either invalid function arguments or side-effects (e.g. IO) associated with allocation. The commit for that change is here:

https://github.com/adamkewley/libdeflater/commit/75dae067ac8abfd7266dda563d234099850c70d2

Amazingly well-spotted on the typos :D, which are now fixed:

https://github.com/adamkewley/libdeflater/commit/b410db129b5ef82aeea606cabd37788fe89d4bfd

As for omitting the `_decompress_ex()` and `crc` functions, it is mostly because this is a first version that has the minimal API required that covers normal use-cases (e.g. loading a whole zlib blob). The `_decompress_ex()` functions feel like they are targeted to more specific use-cases, such as loading a blob you "know" is chunked within some tolerance limit but where you don't know the exact chunk sizes up-front. I will add these methods the next time I get an hour or two to implement the wrappers (+ docs, tests, etc. etc.): again, I think they're useful but weren't the focus of this first version.

On another note, for the benchmarks, they're still not fully-formed yet because there's no comparison between the compression ratio of flate2's default compressors and libdeflate's default compression level, but the results do look promising for this particular corpus (I'll need "bad case" files, such as large files, for comparison)

**Comment by adamkewley** [comment]
Just as an update on this:

- I added the crc32 functions to the API (https://github.com/adamkewley/libdeflater/commit/5b95efde2b681c1669a3b9c02530db3a4baac74d)
- All methods have been renamed to match libdeflate (https://github.com/adamkewley/libdeflater/commit/0d14ddbc6bcd327f066bdee1620bdac2bd9049d5)
- Benchmarks have been updated with Canterbury corpus. Libdeflate is around ~2-3x faster on average (https://github.com/adamkewley/libdeflater/commit/18c8278aff549a80fb5b2b523f90e3d6a21feca8)

My intended next step for this is to throw the bindings out to the wider Rust community for more feedback. If it gains attention, then I'll also add the `_ex` and `adler32` functions as people ask for them. My main goal is to have a working first version of the bindings then adjust as things move forward!
